### DISCUSSION TITLE
Trigger-based statistics dump

### DISCUSSION THREAD

**Discussion by qtt2** [sparcians/map#487] [General]
Hi,

We have some statistics that are calculated at the end of simulation via `simulationTerminating_()`. For example, a `cache.total_read_access` of a cache is the sum of all `cache.bank[0-7].read_access` of its banks 0 to 7. When you use triggers (e.g., start and stop) to dump statistic report for a specific region, this stat does not get updated because `simulationTerminating_()` is not called at trigger points.

I can think of two possible options:

(1) In the report def yaml file, I define something like this

```
cache.bank*.read_access : read_access_$1
```

This prints out the `read_access` per banks, and I need to do a post-processing step to accumulate all `read_access_[0-7]` after the report gets generated. It would be nice if we can define a sum across all instances of `cache.bank*.read_access` in the report.

(2) In the simulator, call a function similar to `simulationTerminating_()` to accumulate partial stats at trigger points. I don't know if something like that already existing in the framework. If so, can someone point me to that?

Thanks,

Tuan

**Comment by klingaard** [comment]
Sorry for the delay -- been out of town.

For the first option, you should be able to create an expression in your report file like we do for Dhrystone MIPS calculation: https://github.com/riscv-software-src/riscv-perf-model/blob/8d187b22e95ac8490489fa3ed9f64a51faf42705/reports/dhry_report.yaml#L14

It's not elegant, but possible.  Another option is to create a [`StatisticDef`](https://sparcians.github.io/map/classsparta_1_1StatisticDef.html) instance in your class that sums them up:
```
   // There are many constructors, using a simpler one
   sparta::StatisticDef cache_bank_sums_{getStatisticSet(), 
                                         "total_read_accesses_in_bank", // stat name
                                         getContainer(),  // stat context -- where the cache.bank*.read_access stats are
                                         R"( cache.bank1.stats.read_access + 
                                             cache.bank2.stats.read_access + 
                                             cache.bank3.stats.read_access` +
                                             cache.bank4.stats.read_access + 
                                             cache.bank5.stats.read_access + 
                                             cache.bank6.stats.read_access + 
                                             cache.bank7.stats.read_access)"};  // expression
```
The neat thing you can do here is to create the `StatisticDef` dynamically, during construction, and create the expression on the fly based on the number of banks.

For (2), if you use the `StatisticDef` class, it will work really well with triggers.  It's only evaluated on triggers.

Hope this helps.
